rollouts/controllers/rollout_controller.go

```diff
@@ -129,35 +129,107 @@ func (r *RolloutReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+	logger := log.FromContext(ctx)
+
+	progressiveStrategy := gitopsv1alpha1.ProgressiveRolloutStrategy{}
+	progressiveStrategy.Spec = gitopsv1alpha1.ProgressiveRolloutStrategySpec{Waves: []gitopsv1alpha1.Wave{}}
+
 	// validate the strategy as early as possible
 	switch typ := rollout.Spec.Strategy.Type; typ {
-	case gitopsv1alpha1.AllAtOnce, gitopsv1alpha1.RollingUpdate:
+	case gitopsv1alpha1.AllAtOnce:
+		wave := gitopsv1alpha1.Wave{MaxConcurrent: math.MaxInt, Targets: rollout.Spec.Targets}
+		progressiveStrategy.Spec.Waves = append(progressiveStrategy.Spec.Waves, wave)
+
+	case gitopsv1alpha1.RollingUpdate:
+		wave := gitopsv1alpha1.Wave{MaxConcurrent: rollout.Spec.Strategy.RollingUpdate.MaxConcurrent, Targets: rollout.Spec.Targets}
+		progressiveStrategy.Spec.Waves = append(progressiveStrategy.Spec.Waves, wave)
+
 	case gitopsv1alpha1.Progressive:
 		strategyRef := types.NamespacedName{
 			Namespace: rollout.Spec.Strategy.Progressive.Namespace,
 			Name:      rollout.Spec.Strategy.Progressive.Name,
 		}
-		if err := r.Get(ctx, strategyRef, progressiveStrategy); err != nil {
+		if err := r.Get(ctx, strategyRef, &progressiveStrategy); err != nil {
 			logger.Error(err, "unable to fetch progressive rollout strategy", "strategyref", strategyRef)
 			// TODO (droot): signal this as a condition in the rollout status
-			return ctrl.Result{}, err
+			return nil, err
 		}
-		logger.Info("progressive rollout is not supported yet, so nothing to reconcile here.")
-		return ctrl.Result{}, nil
+
+		err := r.validateProgressiveRolloutStrategy(ctx, rollout, &progressiveStrategy)
+		if err != nil {
+			logger.Error(err, "progressive rollout strategy failed validation", "strategyref", strategyRef)
+			// TODO (cfry): signal this as a condition in the rollout status
+			return nil, err
+		}
+
 	default:
 		// TODO (droot): signal this as a condition in the rollout status
-		return ctrl.Result{}, fmt.Errorf("%v strategy not supported yet", typ)
+		return nil, fmt.Errorf("%v strategy not supported yet", typ)
 	}
 
-	packageDiscoveryClient := r.getPackageDiscoveryClient(req.NamespacedName)
+	return &progressiveStrategy, nil
+}
 
-	err := r.reconcileRollout(ctx, &rollout, packageDiscoveryClient)
+func (r *RolloutReconciler) validateProgressiveRolloutStrategy(ctx context.Context, rollout *gitopsv1alpha1.Rollout, strategy *gitopsv1alpha1.ProgressiveRolloutStrategy) error {
+	allClusters, err := r.store.ListClusters(ctx, rollout.Spec.Targets.Selector)
 	if err != nil {
-		return ctrl.Result{}, err
+		return err
 	}
 
-	return ctrl.Result{}, nil
+	clustersMap := make(map[string]int)
+	for _, cluster := range allClusters.Items {
+		clustersMap[cluster.Name] = -1
+	}
+
+	for waveIdx, wave := range strategy.Spec.Waves {
+		waveClusters, err := r.store.ListClusters(ctx, rollout.Spec.Targets.Selector, wave.Targets.Selector)
+		if err != nil {
+			return err
+		}
+
+		if len(waveClusters.Items) == 0 {
+			return fmt.Errorf("wave %d does not target any clusters", waveIdx)
```

Comment:
It's a good question. I'll need to do more research around this, but my initial thought is thought is that this should be logged as a warning (versus an error as I do today), still stop the rollout as this should be unexpected and considered a misconfiguration issue, and the validation message shown on the rollout's status. Keeping it as an error at the moment since we still need to iterate on rollout status.